homeassistant/components/surepetcare/__init__.py

```diff
@@ -164,3 +152,64 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+class SurePetcareDataCoordinator(DataUpdateCoordinator):
+    """Handle Surepetcare data."""
+
+    def __init__(self, entry: ConfigEntry, hass: HomeAssistant) -> None:
+        """Initialize the data handler."""
+        self.surepy = Surepy(
+            entry.data[CONF_USERNAME],
+            entry.data[CONF_PASSWORD],
+            auth_token=entry.data[CONF_TOKEN],
+            api_timeout=SURE_API_TIMEOUT,
+            session=async_get_clientsession(hass),
+        )
+        self.lock_states = {
+            LockState.UNLOCKED.name.lower(): self.surepy.sac.unlock,
+            LockState.LOCKED_IN.name.lower(): self.surepy.sac.lock_in,
+            LockState.LOCKED_OUT.name.lower(): self.surepy.sac.lock_out,
+            LockState.LOCKED_ALL.name.lower(): self.surepy.sac.lock,
+        }
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+
+    async def _async_update_data(self) -> dict[int, SurepyEntity]:
+        """Get the latest data from Sure Petcare."""
+        try:
+            return await self.surepy.get_entities(refresh=True)
+        except SurePetcareError as err:
+            raise UpdateFailed(f"Unable to fetch data: {err}") from err
+
+    async def handle_set_lock_state(self, call: ServiceCall) -> None:
+        """Call when setting the lock state."""
+        flap_id = call.data[ATTR_FLAP_ID]
+        state = call.data[ATTR_LOCK_STATE]
+        await self.lock_states[state](flap_id)
+        await self.async_request_refresh()
+
+    def get_pets(self) -> dict[int, str]:
+        """Get pets."""
+        names = {}
+        for surepy_entity in self.data.values():
+            if surepy_entity.type == EntityType.PET and surepy_entity.name:
+                names[surepy_entity.id] = surepy_entity.name
```

Comment:
If we invert the dictionary and have the names be the keys it would be easier to look up the correct pet by name.